core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala

```diff
@@ -315,6 +319,16 @@ private[spark] object CoarseGrainedExecutorBackend extends Logging {
     }
   }
 
+  // Returning T, throwing the exception on failure
+  @scala.annotation.tailrec
```

Comment:
Nit, but a new method tailrec is big overkill here. Just
```for (_ <- 0 until 3) { ... }```

Reply:
Hi @srowen  
This helps the code to be testable, if you want I can change it. But I'd also have to delete the tests. I'll wait for your feedback.

Reply:
Hi @srowen 
I've done as you requested and transformed it into a for loop.